### PR TITLE
Generalization of FSDP common for non-cuda execution

### PR DESCRIPTION
**Pull request by ankurneog**
## Motivation
The FSDP common code for FSDP UT execution is mostly written with cuda device in mind. However other devices such the intel Gaudi supports most of the functionality. We are generalizing the base content so that the UT content can be used for non-cuda device execution.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

**Comment by ankurneog** [comment]
@wconstab  , @albanD  : can you please help with the review . Thanks

**Comment by albanD** [comment]
@wconstab who would have some bandwidth to do these FSDP update to make it work with non-cuda devices?

**Comment by ankurneog** [comment]
@wconstab  : Can you please help with the review and approval, Main changes here : https://github.com/pytorch/pytorch/pull/133209/files#diff-7b5c66f99161fa6a3d9042e80f8c8cc140a64e43445feede46f55e53154f6c3d, others are mostly for adapting to change made in that file

**Comment by ankurneog** [comment]
@albanD  /@wconstab : could you please help with the review and approval. thanks

File: torch/testing/_internal/common_fsdp.py
Comment (wconstab) [inline torch/testing/_internal/common_fsdp.py:98]: nit: s/CUDA/DEVICE/

File: torch/testing/_internal/common_fsdp.py
Comment (wconstab) [inline torch/testing/_internal/common_fsdp.py:654]: just curious, why the difference?

File: torch/testing/_internal/common_fsdp.py
Comment (wconstab) [inline torch/testing/_internal/common_fsdp.py:1121]: ~todo- is this actually unchanged in the new formula?~

as long as `TEST_CUDA := torch.cuda.is_available()`, then `DEVICE_COUNT := torch.cuda.device_count()` so this appears equivalent.

File: torch/testing/_internal/common_fsdp.py
Comment (wconstab) [inline torch/testing/_internal/common_fsdp.py:1186]: why is this left commented?

File: torch/testing/_internal/common_fsdp.py
Comment (wconstab) [inline torch/testing/_internal/common_fsdp.py:1559]: is this a behavior change? were we relying on the 'set_device' for cuda to cover this before and now we make it explicit?

**Review by wconstab — COMMENTED** [review]
mostly LGTM. a couple of small questions and then i would stamp.

Also, fwiw you may want to look into FSDP2 support as well, if you haven't.

**Comment by ankurneog** [comment]
@wconstab  : thank you for your review comments, I have addressed them and also replied to your queries. can you please have a look and help with the approval. thank you.

**Comment by ankurneog** [comment]
@wconstab : gentle reminder , could you kindly do the needful, thanks.

**Comment by ankurneog** [comment]
@wconstab  / @fegin : could you please help with the approval , we have another PR: https://github.com/pytorch/pytorch/pull/135242/  dependent on this. Thank you

**Comment by kwen2501** [comment]
Nice effort.
Curious -- does anyone know the status of FSDP's device support now? @awgu 
Or is the plan to first improve the tests then the library? @ankurneog

**Comment by ankurneog** [comment]
@fegin : could you help with the approval ( there is continuous refactor requited in the fsdp test files after every rebase , due to the change in the common file)

**Comment by ankurneog** [comment]
@wconstab / @fegin  : could you kindly help with the approval and merge. thanks

**Comment by ankurneog** [comment]
@wconstab / @fegin : could you kindly help with the approval and merge.

File: test/distributed/fsdp/test_wrap.py
Comment (kwen2501) [inline test/distributed/fsdp/test_wrap.py:348]: Why do we keep `move_to_cuda`? Does it have an effect or not?

File: torch/testing/_internal/common_fsdp.py
Comment (kwen2501) [inline torch/testing/_internal/common_fsdp.py:81]: `DEVICE_TYPE = "hpu:0"` seems a bit hacky. Can you make it "hpu" and add a `DEVICE_COUNT` field?

Comment (kwen2501) [reply torch/testing/_internal/common_fsdp.py:81]: Another backend has the same "hccl" name.

File: torch/testing/_internal/common_fsdp.py
Comment (kwen2501) [reply torch/testing/_internal/common_fsdp.py:654]: +1, please help us unify code paths to reduce maintenance load.

File: torch/testing/_internal/common_fsdp.py
Comment (kwen2501) [inline torch/testing/_internal/common_fsdp.py:670]: Same here.

File: torch/testing/_internal/common_fsdp.py
Comment (kwen2501) [inline torch/testing/_internal/common_fsdp.py:804]: Same here.

File: torch/testing/_internal/common_fsdp.py
Comment (kwen2501) [inline torch/testing/_internal/common_fsdp.py:1153]: I propose `return DEVICE_COUNT` if there is no specific reason to cap at 8. (I know it is not from this PR, appreciate it if you could give it a try).

**Review by kwen2501 — APPROVED** [review]
LGTM overall. Please see if you can address the comments before landing.

**Comment by ankurneog** [comment]
@kwen2501  :thanks for the review, I have addressed your queries, Could you please help with the merge. I see you unrelated failures, for which i don't permission to bypass

**Comment by ankurneog** [comment]
@albanD , @kwen2501 : Could you please help with the merge . Thank you

**Comment by ankurneog** [comment]
@albanD , @kwen2501 : Could you please help with the merge . The failures look unrelated.

File: torch/testing/_internal/common_fsdp.py
Comment (awgu) [inline torch/testing/_internal/common_fsdp.py:656]: why is HPU sleeping based on `delay_before_reduction_ms` instead of `delay_after_loss_ms`?

**Comment by awgu** [comment]
I am kind of confused why some of the comments left by previous reviewers were left unaddressed or without response.

**Comment by ankurneog** [comment]
> I am kind of confused why some of the comments left by previous reviewers were left unaddressed or without response.

Could you please let me know what needs to be addressed?

File: torch/testing/_internal/common_fsdp.py
Comment (awgu) [reply torch/testing/_internal/common_fsdp.py:656]: this one I just added is new

File: torch/testing/_internal/common_fsdp.py
Comment (awgu) [reply torch/testing/_internal/common_fsdp.py:1557]: @ankurneog this

File: torch/testing/_internal/common_fsdp.py
Comment (awgu) [reply torch/testing/_internal/common_fsdp.py:654]: @ankurneog this

File: torch/testing/_internal/common_fsdp.py
Comment (awgu) [reply torch/testing/_internal/common_fsdp.py:670]: @ankurneog this

File: torch/testing/_internal/common_fsdp.py
Comment (awgu) [reply torch/testing/_internal/common_fsdp.py:804]: @ankurneog this

**Pull request closed without merging**
